src/ensenso.cpp

```diff
@@ -351,85 +501,101 @@ void Ensenso::recordCalibrationPattern(std::string * parameters_dump_info, std::
+
+	return estd::in_place_valid;
 }
 
-Eigen::Isometry3d Ensenso::detectCalibrationPattern(int const samples, bool ignore_calibration)  {
+Result<Eigen::Isometry3d> Ensenso::detectCalibrationPattern(int const samples, bool ignore_calibration)  {
 	discardCalibrationPatterns();
 
 	for (int i = 0; i < samples; ++i) {
 		recordCalibrationPattern();
 	}
 
 	// Disable FlexView (should not be necessary here, but appears to be necessary for cmdEstimatePatternPose)
-	int flex_view = flexView();
-	if (flex_view > 0) setFlexView(0);
+	Result<int> flex_view = flexView();
+	if (!flex_view) return flex_view.error();
+	if (*flex_view > 0) setFlexView(0);
 
 	// Get the pose of the pattern.
 	NxLibCommand command_estimate_pose(cmdEstimatePatternPose);
-	executeNx(command_estimate_pose);
+	Result<void> execute_estimate = executeNx(command_estimate_pose);
+	if (!execute_estimate) return execute_estimate.error().push_description("failed to execute nxlib command");
 
+	// TODO:To this if execute fails as well??
 	// Restore FlexView setting.
-	if (flex_view > 0) {
-		setFlexView(flex_view);
+	if (*flex_view > 0) {
+		setFlexView(*flex_view);
 	}
 
-	Eigen::Isometry3d result = toEigenIsometry(command_estimate_pose.result()["Patterns"][0][itmPatternPose]);
-	result.translation() *= 0.001;
+	Result<Eigen::Isometry3d> eigen_isometry = toEigenIsometry(command_estimate_pose.result()["Patterns"][0][itmPatternPose]);
+	if (!eigen_isometry) return eigen_isometry.error().push_description("failed to detect calibration pattern.");
+
+	eigen_isometry->translation() *= 0.001;
 
 	// Transform back to left stereo lens.
 	if (ignore_calibration) {
-		std::optional<Eigen::Isometry3d> camera_pose = getWorkspaceCalibration();
+		Result<Eigen::Isometry3d> camera_pose = getWorkspaceCalibration();
 		if (camera_pose) {
-			result = *camera_pose * result;
+			*eigen_isometry = *camera_pose * *eigen_isometry;
 		}
 	}
 
-	return result;
+	return *eigen_isometry;
 }
 
 std::string Ensenso::getWorkspaceCalibrationFrame() {
+	// TODO: why not do this?? return getNx<std::string>(stereo_node[itmLink][itmTarget])
 	// Make sure the relevant nxLibItem exists and is non-empty, then return it.
 	NxLibItem item = stereo_node[itmLink][itmTarget];
 	int error;
 	std::string result = item.asString(&error);
 	return error ? "" : result;
 }
 
-std::optional<Eigen::Isometry3d> Ensenso::getWorkspaceCalibration() {
+Result<Eigen::Isometry3d> Ensenso::getWorkspaceCalibration() {
 	// Check if the camera is calibrated.
-	if (getWorkspaceCalibrationFrame().empty()) return std::nullopt;
+	if (getWorkspaceCalibrationFrame().empty()) return estd::error("failed to retrieve workspace calibration.");
 
 	// convert from mm to m
-	Eigen::Isometry3d pose = toEigenIsometry(stereo_node[itmLink]);
-	pose.translation() *= 0.001;
+	Result<Eigen::Isometry3d> pose = toEigenIsometry(stereo_node[itmLink]);
+	if (!pose) return pose.error().push_description("failed to retrieve workspace calibration.");
```

Comment:
```suggestion
	if (!pose) return pose.error().push_description("failed to retrieve workspace calibration");
```

Must be composable :)